### PR TITLE
Pipeline class for pipeline initiation and requirement check

### PR DESCRIPTION
**Pull request by shenshan**

### CODE REVIEW

File: djutils/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+        return msg
+
+    def _check_requirements(self, requirements):
+
+        checked_requirements = {}
+        if self.requried_table_names:
+            for k in self.requried_table_names:
+                if k not in requirements:
+                    raise KeyError(f'Requiring upstream table: {k}')
```
Comment (dimitri-yatsenko) [inline djutils/pipeline.py:35]: f strings are not supported in Python 3.5 and earlier

File: djutils/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    @property
+    def requirements(self):
+
+        if not (self.requried_table_names or self.requried_method_names):
+            return 'No required upstream tables or methods.'
+
+        msg = u'"requirements" needs to be a dictionary with'
+
+        if self.requried_table_names:
+            msg += u' - Keys for upstream tables: {}'.format(self.requried_table_names)
```
Comment (dimitri-yatsenko) [inline djutils/pipeline.py:23]: u strings are not necessary in Python 3

File: djutils/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        self.requried_table_names = required_table_names
+        self.requried_method_names = required_method_names
+
+    @property
+    def requirements(self):
+
+        if not (self.requried_table_names or self.requried_method_names):
+            return 'No required upstream tables or methods.'
+
+        msg = u'"requirements" needs to be a dictionary with'
```
Comment (dimitri-yatsenko) [inline djutils/pipeline.py:20]: u strings are from Python 2

File: djutils/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
+import datajoint as dj
+import inspect
+
+class Pipeline:
```
Comment (dimitri-yatsenko) [inline djutils/pipeline.py:4]: Please provide a docstring describing the purpose and uses of `Pipeline` .

File: djutils/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    def __init__(self, table_classes,
```
Comment (dimitri-yatsenko) [inline djutils/pipeline.py:6]: provide a docstring explaining the arguments

File: djutils/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        if self.requried_method_names:
+            for k in self.requried_method_names:
+                if k not in requirements or not inspect.isfunction(requirements[k]):
+                    raise KeyError(f'Requiring method: {k}')
+                else:
+                    checked_requirements[k] = requirements[k]
+
+        return checked_requirements
+
+    def init_pipeline(self, schema, requirements=None, context={}, add_here=False):
```
Comment (dimitri-yatsenko) [inline djutils/pipeline.py:48]: add docstring explaining parameters

Comment (dimitri-yatsenko) [reply djutils/pipeline.py:48]: Do not use a dictionary as a default parameter value since it's a mutable object.

Comment (eywalker) [reply djutils/pipeline.py:48]: that's a good catch - should be set to `None` and then `context = context or {}` later

File: djutils/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            if requirements:
+                for hook_name, hook_target in requirements.items():
+                    hook_name = '_{}'.format(hook_name)
+                    if hook_name in dir(table_class):
+                        setattr(table_class, hook_name, hook_target)
+
+            print('Initializing {}'.format(table_class.__name__))
+            table = schema(table_class, context=context)
+            context[table.__name__] = table
+            tables[table.__name__] = table
```
Comment (ttngu207) [inline djutils/pipeline.py:86]: here, maybe we can do:
`setattr(self, table.__name__, table)`
and not having to return the `tables`

Users can do the `.` to get the table (as we briefly discussed):
```
subj_pipeline.init_pipeline(
    subj_schema, requirements, add_here=True)
subj_pipeline.Strain
subj_pipeline.Subject
```

**Pull request merged**